### PR TITLE
Vivado-equivalent implementation of Softmax on Quartus

### PR DESCRIPTION
**Pull request by bo3z**
- Implemented `fixed_point_utils.py`, emulating some fixed-point and bit-manipulation operations in Python. These are usually done through HLS, but are needed in Python for generating LUTs. For reasoning behind Softmax LUT, see Vivado implemenation, in `templates/vivado/nnet_utils/nnet_activation.h` .
- Stable and Latency implementation of Softmax. Stable performs well (>98%). Like Vivado, Latency approach needs further work to improve accuracy.
- Expanded PyTest to include Softmax tests on Quartus.

### CODE REVIEW

**Comment by thesps** [comment]
How about moving the computation of the table data out of the writer and into a backend optimizer pass? You could attach the table data to the layer in an analogous way to weights for other layers.

**Comment by bo3z** [comment]
> 

I don't see why not. But then, for consistency sake, we should do that for all of the other activations as well - which, I'm more than happy to do, but feel like this should be a part of a separate PR? Not necessarily softmax-specific?

**Comment by bo3z** [comment]
> It looks essentially like you're using the same type for the address (i) and the data stored at that location (f) by setting f.set_msb_bits(uint_to_binary(i, N))

Re: The logic breakdown behind this is:
1. Given an integer i (the address or index in this case), extract its bits as an array of length N (the number of bits needed for integer i is always less than N, because the setup is such that the index loops from 0 to N); N is usually 10 (table of size 1024)

2. Create a fixed point number, with some number of total bits (default 16) and integer bits (default 6) and set it's first N bits to the bits of i. So in the default case, all of the integer bits (the first 6, take the values of bits of i) and then another 4 take the remaining bits of i. The remaining bits (6 in this case) are unaffected.
3. This is internally converted to float before doing any operation (inv or exp)

To my understanding, this is equivalent to the Vivado C++ function below:
```
template<class data_T, typename CONFIG_T>
inline float softmax_real_val_from_idx(unsigned i){
    // Treat the index as the top N bits
    static constexpr int N = ceillog2(CONFIG_T::table_size); // number of address bits for table
    data_T x(0);
    x(x.width-1, x.width-N) = i;
    return (float) x;
} 
```
I did also check the outputs with Vivado HLS and a stand-alone Python script to test this functionality, and the values were equal.

4. Final step, given a float, obtained by such bit manipulation (steps 2-3), do the function. This is done in Python, by the two functions, `exp_float` and `inv_float`. However, this produces a floating point result, which is saved to an external file. This is fine, because Quartus HLS can inherently handle floating to fixed point conversion at run-time.

**Comment by thesps** [comment]
> 4\. the

I think the issue is just that, in the Vivado code you showed, the computation is done for `data_T` (the data type of the layer input), whereas the new Quartus writer code is doing this on the `exp_table_t` and `inv_table_t`. But in general those three could all be different.

File: hls4ml/writer/quartus_writer.py [hunk 1]
```diff
@@ -597,10 +599,30 @@ def __write_exp_table(self, model, path):
...
+            if layer.name == 'softmax':
+                ac_type = model.config.get_layer_config_value(layer, 'exp_table_t')
+                if ac_type is not None:
+                    re_search = r'ac_fixed<(\d+),\s?(\d+),\s?(true|false)'
+                    match = re.search(re_search, ac_type)
+                    if match is not None:
+                        fp_bits = int(match.group(1))
+                        fp_integer = int(match.group(2))
+                        fp_signed = False if match.group(2)=='false' else True
+
```
[the remark this answers is not in the document]
Comment (thesps) [reply hls4ml/writer/quartus_writer.py:619]: In the config it's a string, but it should already have been converted to one of [these](https://github.com/fastmachinelearning/hls4ml/blob/master/hls4ml/model/types.py#L49) `PrecisionType` objects and its derived classes, and attached as an attribute of the `Layer` (not the config).

**Comment by bo3z** [comment]
> I think the issue is just that, in the Vivado code you showed, the computation is done for `data_T` (the data type of the layer input), whereas the new Quartus writer code is doing this on the `exp_table_t` and `inv_table_t`. But in general those three could all be different.

Exactly, I just realised that an hour ago. So it needs to be changed from  `exp_table_t` and `inv_table_t` to the precision used by that layer.

File: hls4ml/writer/quartus_writer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bo3z) [reply hls4ml/writer/quartus_writer.py:619]: This is correct, RegEx was removed. Wasn't aware that attributes store this information.

**Comment by bo3z** [comment]
Addressed discrepancy between data and tables, as explained above. Manually checked results to Keras prediction for 5 tests, on 5 data points. Results are accurate. RegEx was removed, in favour of attributes. Commit with main changes here: https://github.com/fastmachinelearning/hls4ml/pull/540/commits/937cc493835211c9e896924ef26dc523f1103cbd

File: hls4ml/writer/quartus_writer.py [hunk 2]
```diff
@@ -596,11 +621,33 @@ def __write_exp_table(self, model, path):
...
+        # 6 bits for integer part, 10 bits for decimal - total, 16
+        fp_bits = 16
+        fp_integer = 6
+        fp_signed = True
+
+        # Exp table should use the same precision as exp_table, as seen in Vivado code
+        # init_exp_table<data_T, CONFIG_T>(exp_table);
+        for layer in model.get_layers():
+            if layer.name == 'softmax':
+                ac_type = layer.get_attr('accum_t')
```
Comment (bo3z) [inline hls4ml/writer/quartus_writer.py:635]: I couldn't find an attribute for `input_t` in the softmax layer, only manage to find `result_t` and `accum_t`. Is `accum_t` equivalent to `input_t` in this case?

Comment (bo3z) [reply hls4ml/writer/quartus_writer.py]: Sorted, changed to input
